modules/core/src/main/java/org/apache/ignite/internal/processors/service/GridServiceProcessor.java

```diff
@@ -610,178 +568,95 @@ private PreparedConfigurations prepareServiceConfigurations(Collection<ServiceCo
-                    try (Transaction tx = serviceCache().txStart(PESSIMISTIC, READ_COMMITTED)) {
-                        for (ServiceConfiguration cfg : cfgsCp) {
-                            try {
-                                writeServiceToCache(res, cfg);
-                            }
-                            catch (IgniteCheckedException e) {
-                                if (X.hasCause(e, ClusterTopologyCheckedException.class))
-                                    throw e; // Retry.
-                                else
-                                    U.error(log, e.getMessage());
-                            }
-                        }
-
-                        tx.commit();
-                    }
-                }
-
-                break;
+            if (ctx.isStopping()) {
+                return new GridFinishedFuture<>(new IgniteCheckedException("Failed to deploy services, " +
+                    "node is stopping: " + cfgs));
             }
-            catch (IgniteException | IgniteCheckedException e) {
-                for (String name : res.servicesToRollback())
-                    depFuts.remove(name).onDone(e);
 
-                if (X.hasCause(e, ClusterTopologyCheckedException.class)) {
-                    if (log.isDebugEnabled())
-                        log.debug("Topology changed while deploying services (will retry): " + e.getMessage());
-                }
-                else {
-                    res.onDone(new IgniteCheckedException(
-                        new ServiceDeploymentException("Failed to deploy provided services.", e, cfgs)));
+            if (cfgs.isEmpty())
+                return new GridFinishedFuture<>();
 
-                    return res;
-                }
-            }
-            finally {
-                if (ctx.deploy().enabled())
-                    ctx.cache().context().deploy().ignoreOwnership(false);
-            }
-        }
+            PreparedConfigurations srvCfg = prepareServiceConfigurations(cfgs, dfltNodeFilter, false);
 
-        if (ctx.clientDisconnected()) {
-            IgniteClientDisconnectedCheckedException err =
-                new IgniteClientDisconnectedCheckedException(ctx.cluster().clientReconnectFuture(),
-                    "Failed to deploy services, client node disconnected: " + cfgs);
+            List<ServiceConfiguration> cfgsCp = srvCfg.cfgs;
 
-            for (String name : res.servicesToRollback()) {
-                GridServiceDeploymentFuture fut = depFuts.remove(name);
+            List<GridServiceDeploymentFuture> failedFuts = srvCfg.failedFuts;
 
-                if (fut != null)
-                    fut.onDone(err);
-            }
+            GridServiceDeploymentCompoundFuture res = new GridServiceDeploymentCompoundFuture();
 
-            return new GridFinishedFuture<>(err);
-        }
+            if (!cfgsCp.isEmpty()) {
+                cfgsCp.sort(Comparator.comparing(ServiceConfiguration::getName));
```

Comment:
There is no point in sorting the configurations. We did it before to avoid deadlocks. This is not needed now, since we don't run transactions, so there cannot be any deadlocks.

Reply:
Removed.